### PR TITLE
自动加载注册model目录及其子目录中中所有数据模型

### PR DESCRIPTION
**Pull request by roy2an**
##### Checklist

- [x] `npm test` passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows commit guidelines

##### Description of change

<pre>
<code>
model
|-monkey.js
|-Person.js
|-other.js
|-post.js
|-user.js
</code>
</pre>
将原有model文件夹改为如下结构
<pre>
<code>
model
|-subfolder
    |-monkey.js
    |-Person.js
|-other.js
|-post.js
|-user.js
</code>
</pre>
会有如下错误
<pre>
<code>
Base
      ✓ sequelize init success
      1) ctx model property getter
      ✓ model not load non Sequelize files
      2) has right tableName
    Database options
      ✓ should work with default config
      ✓ should work with fixture configs
    Test model
      3) User.test method work
      ✓ should work timestramp
    Test controller
      ✓ should get data from create (39ms)
    Associate
      ✓ ctx model associate init success

  7 passing (1s)
  3 failing

1) test/plugin.test.js
       Base
         ctx model property getter:
     AssertionError [ERR_ASSERTION]:   # test/plugin.test.js:29

  assert.ok(ctx.model.Monkey)
            |   |     |
            |   |     undefined
            |   Sequelize{options:#Object#,config:#Object#,dialect:#MysqlDialect#,queryInterface:#QueryInterface#,models:#Object#,modelManager:#ModelManager#,connectionManager:#ConnectionManager#,importCache:#Object#,test:#Object#,Post:#function#,User:#function#}
            Object{request:#Object#,response:#Object#,app:#MeesengerApplication#,req:#Object#,res:#ServerResponse#,onerror:#function#,originalUrl:"/",starttime:1527234130183}

      at Context.it (test/plugin.test.js:29:14)
      [use `--full-trace` to display the full stack trace]

  2) test/plugin.test.js
       Base
         has right tableName:
     TypeError: Cannot read property 'tableName' of undefined
      at Context.it (test/plugin.test.js:41:14)
      [use `--full-trace` to display the full stack trace]

  3) test/plugin.test.js
       Test model
         User.test method work:
     TypeError: Cannot read property 'create' of undefined
      at Function.User.test (test/fixtures/apps/model-app/app/model/user.js:21:43)
      at User.test.next (<anonymous>)
      at new Promise (<anonymous>)
      at new Promise (<anonymous>)
      [use `--full-trace` to display the full stack trace]
</code>
</pre>
此改动可以修复这个问题

### CODE REVIEW

File: lib/loader.js [hunk 1]
```diff
@@ -52,25 +52,34 @@ function loadModel(app) {
...
+}
 
-  for (const name of Object.keys(app[MODELS])) {
-    const klass = app[MODELS][name];
-
-    if ('associate' in klass) {
-      klass.associate();
+function recursionAssociate(values) {
+  for (const name of Object.keys(values)) {
+    const kclass = values[name];
```
Comment (huacnlee) [inline lib/loader.js]: kclass -> klass

File: lib/loader.js [hunk 2]
```diff
@@ -52,25 +52,34 @@ function loadModel(app) {
...
+    const klass = values[name];
     if ('sequelize' in klass) {
       app.model[name] = klass;
 
       if ('classMethods' in klass.options || 'instanceMethods' in klass.options) {
         app.logger.error(`${name} model has classMethods/instanceMethods, but it was removed supports in Sequelize V4.\
 see: http://docs.sequelizejs.com/manual/tutorial/models-definition.html#expansion-of-models`);
       }
+    } else if(klass){
+      recursionModel(klass, app);
```
Comment (huacnlee) [inline lib/loader.js]: 没看懂产出的效果

举例说明一下

Comment (roy2an) [reply lib/loader.js]: 项目中很有可能出现大量数据模型的情况，如果都直接放在model文件夹下略显杂乱。虽然原本的方法会遍历读取所有子目录中的文件，遇到的问题是数据表虽然同步了，但其实model没有被注册到app.model中，associate方法也没有被调用。随意修改了loader.js文件

Comment (huacnlee) [reply lib/loader.js]: 我的意思是这样改动会影响的结果你需要详细描述出来

**Comment by huacnlee** [comment]
package-lock.json 去掉

其他的没问题了

File: lib/loader.js [hunk 3]
```diff
@@ -52,25 +52,33 @@ function loadModel(app) {
...
+  for (const name of Object.keys(values)) {
+    const klass = values[name];
     if ('sequelize' in klass) {
       app.model[name] = klass;
 
       if ('classMethods' in klass.options || 'instanceMethods' in klass.options) {
         app.logger.error(`${name} model has classMethods/instanceMethods, but it was removed supports in Sequelize V4.\
 see: http://docs.sequelizejs.com/manual/tutorial/models-definition.html#expansion-of-models`);
       }
+    } else if (klass) {
```
Comment (huacnlee) [inline lib/loader.js]: 这里我没看懂, 为何是 `else if (klass)`, `klass` 在这里是一定会有值的。

我的意思是语义上有歧议。看不出来是为了递归加载子目录的

Comment (roy2an) [reply lib/loader.js]: 我修改一下

**Comment by huacnlee** [comment]
在没有更好的方案之前，我建议你将子文件夹里面的移动出来。
实际你举的例子来看，是没有必要拆分文件夹的

**Pull request closed without merging**